File: code_files_R/LikelihoodFunc_y.R
# Likelihood function for Output Gap (y)
# Model: dy_t = -λ_y * y_t * Δ + σ_y * sqrt(Δ) * ε_t
# where ε_t ~ N(0,1)

LikelihoodFunc_y <- function(outputgap, param, Delta = 1/12) {
  
  # Number of observations
  n <- length(outputgap)
  
  # Parameters: sigma_y and lambda_y
  sigmay  <- abs(param[1])   # volatility (ensure positive)
  lambday <- param[2]        # mean reversion speed
  
  # Demean output gap (center around zero)
  y <- outputgap - mean(outputgap)
  
  # Initialize storage
  eps <- numeric(n - 1)
  vareps <- numeric(n - 1)
  logLik <- numeric(n - 1)
  stdeps <- numeric(n - 1)
  
  # Loop over time
  for (i in 1:(n - 1)) {
    # Ornstein-Uhlenbeck discrete approximation
    eps[i] <- y[i + 1] - exp(-lambday * Delta) * y[i]
    vareps[i] <- (sigmay^2 / (2 * lambday)) * (1 - exp(-2 * lambday * Delta))
    
    # standardized residuals
    stdeps[i] <- eps[i] / sqrt(vareps[i])
    
    # log-likelihood contribution (Normal)
    Nv <- 1
    logLik[i] <- log(1 / ((2 * pi)^(Nv / 2) * sqrt(vareps[i]))) -
      0.5 * (eps[i]^2 / vareps[i])
  }
  
  # Sum of log-likelihood (negative for minimization)
  sumloglik <- -sum(logLik[-1])  # skip first obs (like MATLAB)
  
  # Return as a list (for flexibility)
  return(list(
    sumloglik = sumloglik,
    logLik = logLik,
    stdeps = stdeps
  ))
}

File: code_files_MATLAB/LikelihoodFuncy.m
function [sumloglik,logLik,stdeps]=LikelihoodFuncy(outputgap,param)

global Delta

n=length(outputgap);

sigmay=abs(param(1,1));
lambday=param(2,1); 
y=outputgap-mean(outputgap);


for i=1:n-1,   
    
    %Approximate solution of Ornstein-Uhlenbeck (OU)
    eps(i,1)=y(i+1,1)-exp(-lambday*Delta)*y(i,1);
    vareps=sigmay^2/(2*lambday)*(1-exp(-2*lambday*Delta));
    stdeps(i,1)=eps(i,1)/sqrt(vareps);
    
    % normal likelihood for whole time period
    epsVec=[eps(i,1)];
    VarMat=[vareps];
    Nv=length(epsVec);

    logLik(i,1)=log(1./((2.*pi)^(Nv/2).*sqrt(det(VarMat))))-1/2*(epsVec*(VarMat\epsVec'));
end

sumloglik=-sum(logLik(2:end,1)); % this the sum of log likelihood (in negative sign since fminsearch minimizes)

end